namui/src/namui/draw/text.rs

```diff
@@ -1,36 +1,135 @@
+    let left = get_left_in_align(command.x as f32, command.align, total_width);
 
-    let metrics = font.get_metrics();
+    let mut bottom_of_fonts: HashMap<String, f32> = HashMap::new();
 
-    let bottom = command.y as f32 + get_bottom_of_baseline(&command.baseline, &metrics);
+    let mut x = left;
 
-    let left = get_left_in_align(command.x as f32, command.align, width);
+    for GlyphGroup {
+        glyph_ids,
+        width,
+        end_index: _,
+        font,
+    } in glyph_groups
+    {
+        let bottom = bottom_of_fonts
+            .get(&font.id)
+            .map(|bottom| bottom + font.size as f32)
+            .unwrap_or_else(|| {
+                let metrics = font.get_metrics();
+                let bottom = command.y as f32 + get_bottom_of_baseline(&command.baseline, &metrics);
+                bottom_of_fonts.insert(font.id.clone(), bottom);
+                bottom
+            });
 
-    let text_blob = TextBlob::from_text(&command.text, font.borrow());
+        let text_blob = TextBlob::from_glyph_ids(&glyph_ids, &font);
 
-    namui_context
-        .surface
-        .canvas()
-        .draw_text_blob(&text_blob, left, bottom, &paint);
+        namui_context
+            .surface
+            .canvas()
+            .draw_text_blob(&text_blob, x, bottom, &paint);
+
+        x += width;
+    }
+}
+
+#[derive(Debug)]
+struct GlyphGroup {
+    glyph_ids: Vec<u16>,
+    end_index: usize,
+    width: f32,
+    font: Arc<Font>,
+}
+fn get_glyph_groups(
+    text: &str,
+    fonts: impl Iterator<Item = Arc<Font>>,
+    paint: &Arc<Paint>,
+) -> Vec<GlyphGroup> {
+    let mut groups: Vec<GlyphGroup> = vec![];
+    let mut non_calculated_char_and_indexes: Vec<(char, usize)> = text
+        .chars()
+        .enumerate()
+        .map(|(index, char)| (char, index))
+        .collect();
+    let mut fonts = fonts.peekable();
+
+    while !non_calculated_char_and_indexes.is_empty() && fonts.peek().is_some() {
+        let font = fonts.next().unwrap();
+
+        let text = non_calculated_char_and_indexes
+            .iter()
+            .map(|(char, _)| char)
+            .collect::<String>();
+
+        let glyph_ids = font.get_glyph_ids(&text);
+
+        let mut available_glyph_id_and_indexes = vec![];
+        for (index, glyph_id) in glyph_ids.iter().enumerate() {
+            if *glyph_id != 0 {
+                available_glyph_id_and_indexes.push((*glyph_id, index));
+                non_calculated_char_and_indexes.retain(|(_, index2)| index != *index2);
+            }
+        }
+
+        if available_glyph_id_and_indexes.is_empty() {
+            continue;
+        }
+
+        let available_glyph_id_and_index_and_width: Vec<(u16, usize, f32)> = {
+            let available_glyph_ids: Vec<_> = available_glyph_id_and_indexes
+                .iter()
+                .map(|(glyph_id, _)| *glyph_id)
+                .collect();
+
+            let widths = font.get_glyph_widths(&available_glyph_ids, Option::Some(paint));
```

Comment:
It's because I want to call external call(`font.get_glyph_widths`) only once per font.